### PR TITLE
Basic entity design

### PR DESCRIPTION
**Pull request by gdlcf88**

### CODE REVIEW

**Comment by RayMMond** [comment]
Suggestion: `Asset` needs a property `int Priority`.
In some cases, a customer is allowed to book a Room Type(`AssetCategory`), such as Premier/Suite, but not allowed to book one specific room(`Asset`), thus the Booking module needs to select **one** `Asset` from all `Assets` in `AssetCategory`, and user should be able to configure priority of each `Asset` during the selection process.

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+{
+    [NotNull]
+    public string Name { get; }
+    
+    /// <summary>
+    /// This property determines whether assets can be occupied by default when there is no schedule created.
+    /// The SchedulePolicy of <see cref="AssetCategory"/> is preferred.
+    /// </summary>
+    public virtual AssetSchedulePolicy DefaultSchedulePolicy { get; }
+
```
Comment (RayMMond) [inline src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs:17]: Feature suggestion: **Advance Booking Days**. Users should be able to configure the days before assets become available. For example, a hotel room at 10th July can be available when 1st July.
The property `AdvanceBookingDays` in `AssetDefinition` determines the days before this asset became available by default.

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetCategories/AssetCategory.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    public virtual Guid? PeriodSchemeId { get; protected set; }
+    
+    /// <summary>
+    /// This property determines whether assets can be occupied by default when there is no schedule created.
+    /// The SchedulePolicy of <see cref="Asset"/> is preferred.
+    /// This value only affects assets in this category, but not assets in children.
+    /// Will fall back to <see cref="AssetDefinition"/> if the value here is <c>null</c>.
+    /// </summary>
+    public virtual AssetSchedulePolicy? DefaultSchedulePolicy { get; protected set; }
+
```
Comment (RayMMond) [inline src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetCategories/AssetCategory.cs:38]: The `AssetCategory` should have `AdvanceBookingDays`  property too. When it is null, fall back to `AssetDefinition`.

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetSchedules/AssetSchedule.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    public virtual TimeSpan StartingTime { get; protected set; }
+    
+    public virtual TimeSpan Duration { get; protected set; }
+    
+    /// <summary>
+    /// Accept or reject occupying within this time frame.
+    /// </summary>
+    public virtual AssetSchedulePolicy SchedulePolicy { get; protected set; }
+
```
Comment (RayMMond) [inline src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetSchedules/AssetSchedule.cs:23]: The `AssetSchedule` should have `AdvanceBookingDays`  property too.
When it is null, fall back to `AssetCategory`.

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/Assets/Asset.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    /// Will fall back to <see cref="AssetCategory"/> if the value here is <c>null</c>.
+    /// </summary>
+    public virtual Guid? PeriodSchemeId { get; protected set; }
+    
+    /// <summary>
+    /// This property determines whether assets can be occupied by default when there is no schedule created.
+    /// Will fall back to <see cref="AssetCategory"/> if the value here is <c>null</c>.
+    /// </summary>
+    public virtual AssetSchedulePolicy? DefaultSchedulePolicy { get; protected set; }
+
```
Comment (RayMMond) [inline src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/Assets/Asset.cs:32]: The `Asset` should have `AdvanceBookingDays`  property too.
When it is null, fall back to `AssetCategory` or 'AssetSchedule'.

**Review by RayMMond — CHANGES_REQUESTED** [review]

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs (hunk 1 above)
Comment (gdlcf88) [reply src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs:17]: Thanks. I will do this change and use the name `DaysInAdvance` (AdvanceBookingDays).

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/Assets/Asset.cs (hunk 4 above)
Comment (gdlcf88) [reply src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/Assets/Asset.cs:32]: AssetSchedule means "A day of an asset".

So I think the fallback ordering is:
AssetSchedule > Asset > AssetCategory > AssetDefinition

**Comment by gdlcf88** [comment]
@RayMMond - I created a new aggregate root named `AssetPeriodScheme` since we need it to specify a PeriodScheme for a date of an Asset.

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs (hunk 1 above)
Comment (RayMMond) [reply src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs:23]: maybe `AvailableDaysInAdvance` is better?

**Review by RayMMond — APPROVED** [review]
LGTM

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs (hunk 1 above)
Comment (gdlcf88) [reply src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs:23]: I think "Available" is not needed, `DaysInAdvance` is clear and enough, or do you have any other better name?

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    /// This property determines whether assets can be occupied by default when there is no schedule created.
+    /// The property value from <see cref="AssetCategory"/> is preferred.
+    /// </summary>
+    public virtual AssetSchedulePolicy DefaultSchedulePolicy { get; }
+    
+    /// <summary>
+    /// How many days in advance can you occupy the asset.
+    /// The property value from <see cref="AssetCategory"/> is preferred.
+    /// </summary>
+    public virtual int DaysInAdvance { get; protected set; }
```
Comment (RayMMond) [inline src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs:22]: `protected set`  needed?

**Review by RayMMond — CHANGES_REQUESTED** [review]

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs (hunk 5 above)
Comment (gdlcf88) [reply src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs:22]: https://github.com/abpframework/abp/blob/dev/modules/blogging/src/Volo.Blogging.Domain/Volo/Blogging/Comments/Comment.cs#L10-L14

I am not sure whether a property should always have a protected setter, but doing it is safer for now.

Comment (gdlcf88) [reply src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs:22]: Sorry, I was wrong. This `AssetDefinition` is not an entity class. The setter can be removed.

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs (hunk 1 above)
Comment (yinchang0626) [reply src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetDefinitions/AssetDefinition.cs:23]: please consider that  `Asset` takes some days to prepare

**Review by yinchang0626 — APPROVED** [review]

**Review by RayMMond — APPROVED** [review]

**Review by yinchang0626 — APPROVED** [review]

**Pull request merged**